File: javascript/medium/coinChange.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    const memo = new Array();
    for(let i = 0; i <= amount; i++) {
        memo[i] = Number.MAX_SAFE_INTEGER;
    }
    memo[0] = 0;
    for(let coin of coins) {
        for(let i = 0; i < memo.length; i++) {
            if(i >= coin && memo[i - coin] !== Number.MAX_SAFE_INTEGER) {
                memo[i] = Math.min(memo[i], (memo[i - coin] + 1));
            }
        }
    }
    return (memo[memo.length-1] === Number.MAX_SAFE_INTEGER) ? -1 : memo[memo.length-1];
};

File: ruby/medium/coin_change.rb
# @param {Integer[]} coins
# @param {Integer} amount
# @return {Integer}
MAX = (2 ** 31) - 1
def coin_change(coins, amount)
    memo = Array.new(amount+1).fill(MAX)
    memo[0] = 0
    coins.each do |coin|
        for i in 1..amount
            if i >= coin && memo[i-coin] != MAX
                memo[i] = [memo[i], (memo[i-coin] + 1)].min
            end
        end
    end
    return memo.last == MAX ? -1 : memo.last
end
